File: src/lib/protocol/event-verify.ts
import { verifySchnorrHex } from './crypto';
import { computeEventId } from './event-id';
import type { NostrEvent } from './event';

export type VerificationResult =
  | { ok: true; event: NostrEvent }
  | { ok: false; code: 'id_mismatch' | 'bad_signature'; message: string };

export function verifyEvent(event: NostrEvent): VerificationResult {
  const id = computeEventId(event);
  if (id !== event.id)
    return {
      ok: false,
      code: 'id_mismatch',
      message: 'event id does not match payload',
    };
  const valid = verifySchnorrHex(event.sig, event.id, event.pubkey);
  if (!valid)
    return {
      ok: false,
      code: 'bad_signature',
      message: 'event signature is invalid',
    };
  return { ok: true, event };
}

File: crates/lkjstr-protocol/src/event_verify.rs
use crate::{NostrEvent, UnsignedNostrEvent, compute_event_id, verify_schnorr_hex};

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum VerificationCode {
    IdMismatch,
    BadSignature,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum VerificationResult {
    Ok(NostrEvent),
    Err {
        code: VerificationCode,
        message: String,
    },
}

pub fn verify_event(event: &NostrEvent) -> VerificationResult {
    let unsigned = UnsignedNostrEvent {
        pubkey: event.pubkey.clone(),
        created_at: event.created_at,
        kind: event.kind,
        tags: event.tags.clone(),
        content: event.content.clone(),
    };
    let Ok(id) = compute_event_id(&unsigned) else {
        return VerificationResult::Err {
            code: VerificationCode::IdMismatch,
            message: "event id does not match payload".to_owned(),
        };
    };
    if id != event.id {
        return VerificationResult::Err {
            code: VerificationCode::IdMismatch,
            message: "event id does not match payload".to_owned(),
        };
    }
    if !verify_schnorr_hex(&event.sig, &event.id, &event.pubkey) {
        return VerificationResult::Err {
            code: VerificationCode::BadSignature,
            message: "event signature is invalid".to_owned(),
        };
    }
    VerificationResult::Ok(event.clone())
}
